### PR TITLE
new layout type `flex-item`

### LINKED ISSUE #2778: Intent to implement: new layout type `flex-item`
A new layout type will be called `flex-item`. It will be similar to the `fill` layout, but instead of trying to take all of the space of the container, it will take the available space based on the flexbox model.

### Declaration
```
<amp-element layout=flex-item></amp-element>
```

### Implementation
```
.-amp-layout-flex-item {
  flex: 1 1 0;
  align-self: stretch;
  height: 100%;
  width: 100%;
}
```

Notice that `height=100%` and `width=100%` are overridable defaults and are only there to ensure that the container indeed specifies `display: flex`. Alternatively, the element can force `display: flex` on the parent.

### Motivation
The size-defining layouts currently include `fixed`, `responsive` and `fill`. The main goal of these layouts is to ensure that the element's size does NOT change when resources are loaded, which would otherwise lead to constant reflows and a broken user experience.

The `flex-item` layout is similar - it ensures that the element is not sporadically resized while also giving it a more flexible positioning options. It'd also allow the element to participate in the flexbox layouts.

### Workaround
Currently the same can be arranged using the following workaround:
```
<style>
.container {
  display: flex;
}
.childN {
  position: relative;
  flex: 1 1 0;
}
</style>

<div class="container">
  <div class="child1">...</div>
  <div class="child2">...</div>
  <div class="childN">
    <amp-element layout="fill"></amp-element>
  </div>
</div>
```

The proposed `layout=flex-item` will allow the same but without the need of, otherwise meaningless, `div.childN`.

### PR DESCRIPTION
**Pull request by zhouyx**
Intent to implement: new layout type `flex`

Closes #2778.

### CODE REVIEW

File: src/custom-element.js [hunk 1]
```diff
@@ -231,6 +231,9 @@ export function applyLayout_(element) {
     // Do nothing. Elements themselves will check whether the supplied
     // layout value is acceptable. In particular container is only OK
     // sometimes.
+  } else if (layout == Layout.FLEX) {
+    //Display the element as flex
```
Comment (dvoytenko) [inline src/custom-element.js]: nit: space

File: src/custom-element.js [hunk 2]
```diff
@@ -231,6 +231,9 @@ export function applyLayout_(element) {
...
+    element.style.display = 'flex';
```
Comment (dvoytenko) [inline src/custom-element.js]: Element itself cannot have `display: flex` - only parent must have it. So, basically we don't need to do this here. This might the reason - hard to say.

File: examples/everything.amp.html [hunk 3]
```diff
@@ -136,7 +136,10 @@ <h1 id="top">AMP #0</h1>
     <a class="goto" href="#img0">Image0</a>
     <a class="goto" href="#footer">Footer</a>
   </p>
-
+  
+  <!--Add to test flex layout-->
+  <amp- src="https://lh3.googleusercontent.com/5rcQ32ml8E5ONp9f9-Rf78IofLb9QjS5_0mqsY1zEFc=w400-h300-no-n" width=400 height=300 layout=flex></amp-img>
```
Comment (dvoytenko) [inline examples/everything.amp.html]: Without `display: flex` parent this image should not display. I think that's what you are running into. Please go over https://css-tricks.com/snippets/css/a-guide-to-flexbox/ - there are some good examples here.

**Comment by cramforce** [comment]
Please update the title, since this is no longer an `Intent`. Something like "Introduces a new layout type `flex`"

And then please edit the description to reference the respective issue number.

**Comment by cramforce** [comment]
Please see if you can add a few unit tests. Since the logic is simple, the tests should also be pretty simple.

**Comment by erwinmombay** [comment]
a couple of tasks to complete
 
- [ ] update docs
- [ ] add demo/sample

File: examples/layout-flex-item.amp.html [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      height: 500px;
+    }
+    .container-col {
+      display: flex;
+      flex-direction: column;
+      height: 900px;
+    }
+    </style>
+    
+    <script async src="../../dist/amp.js"></script>
```
Comment (erwinmombay) [inline examples/layout-flex-item.amp.html]: switch to v0.js, also make sure this file is a valid amp page

File: src/custom-element.js [hunk 5]
```diff
@@ -241,7 +243,7 @@ export function applyLayout_(element) {
  * @param {!Node} node
  * @return {boolean}
  */
-function isInternalOrServiceNode(node) {
+function isInternalOrServiceNode(node) {
```
Comment (erwinmombay) [inline src/custom-element.js]: remove space

**Comment by zhouyx** [comment]
@cramforce PTAL

File: src/custom-element.js [hunk 6]
```diff
@@ -231,6 +231,8 @@ export function applyLayout_(element) {
     // Do nothing. Elements themselves will check whether the supplied
     // layout value is acceptable. In particular container is only OK
     // sometimes.
+  } else if (layout == Layout.FLEX_ITEM) {
+    //Do nothing.
```
Comment (cramforce) [inline src/custom-element.js]: nit: Space after `//`

Comment (cramforce) [inline src/custom-element.js]: Please expand on comment.

**Comment by cramforce** [comment]
Looks great. We will need an associated validator change. Please file an issue for that.

**Comment by cramforce** [comment]
AH, shoot. The codegen for es6 modules is bad. Make an object per module.

**Comment by erwinmombay** [comment]
@cramforce wrong PR?

**Comment by cramforce** [comment]
Some ways forward, but one would be the big guns: `setCollapseProperties` and `setCollapseObjectLiterals`. Start with the former. If this works it would be big! It might work, because it only touches generated code for ES6 modules that should be super clean.

**Comment by cramforce** [comment]
@erwinmombay LOL yeah

File: spec/amp-html-layout.md [hunk 7]
```diff
@@ -68,6 +68,9 @@ the layout of a `fill` element matches its parent.
 
 The element lets its children to define its size, much like a normal HTML `div`.
 
+### `flex-item`
+The element and other elements with layout type `flex-item` take their parent container's available space. 
+
```
Comment (erwinmombay) [inline spec/amp-html-layout.md]: maybe add a at the end, "when parent has `display: flex`"

File: examples/layout-flex-item.amp.html [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  <amp-img id="child6" src="https://lh3.googleusercontent.com/Z4gtm5Bkxyv21Z2PtbTf95Clb9AE4VTR6olbBKYrenM=w400-h300-no-n" layout=flex-item></amp-img>
+</div>
+</body>
+</html>
+
+<h3>Element without layout typle 'flex-item', have to set children's height and width manually</h3>
+<div class="container">
+   <amp-img id="child7" src="https://lh3.googleusercontent.com/pSECrJ82R7-AqeBCOEPGPM9iG9OEIQ_QXcbubWIOdkY=w400-h300-no-n" height = '100px' width = '150px'></amp-img>
+  <amp-img id="child8" src="https://lh3.googleusercontent.com/5rcQ32ml8E5ONp9f9-Rf78IofLb9QjS5_0mqsY1zEFc=w400-h300-no-n" height = '100px' width = '150px'></amp-img>
+  <amp-img id="child9" src="https://lh3.googleusercontent.com/Z4gtm5Bkxyv21Z2PtbTf95Clb9AE4VTR6olbBKYrenM=w400-h300-no-n" height = '100px' width = '150px'></amp-img>
```
Comment (dvoytenko) [inline examples/layout-flex-item.amp.html]: Please only use `"` (double quotes) for HTML attributes and no spaces around `=`. I.e. `height = '100px'` -> `height="100px"`

File: spec/amp-html-layout.md [hunk 9]
```diff
@@ -68,6 +68,8 @@ the layout of a `fill` element matches its parent.
 
 The element lets its children to define its size, much like a normal HTML `div`.
 
+### `flex-item`
+The element and other elements with layout type `flex-item` take their parent container's available space when parent has display : flex.
```
Comment (dvoytenko) [inline spec/amp-html-layout.md]: We need to rephrase "container's available space" to something like "container's remaining space".

File: spec/amp-html-layout.md [hunk 10]
```diff
@@ -104,6 +106,7 @@ maintaining the height based on the aspect ratio.
 container. Its children are rendered immediately.
 - `nodisplay`: The component takes up zero space on the screen as if its display style was `none`.
 The `width` and `height` attributes are not required.
+- `flex-item`: Element size will be determined by the parent element and the number of other elements inside parent.
```
Comment (dvoytenko) [inline spec/amp-html-layout.md]: "according to `display:flex` CSS layout."

File: css/amp.css [hunk 11]
```diff
@@ -78,6 +78,14 @@ body {
   right: 0;
 }
 
+.-amp-layout-flex-item {
```
Comment (dvoytenko) [inline css/amp.css]: Let's add here `display: block` since you are documenting it explicitly below. I think this is what actually happens by default, but would be nicer to be explicit, I think.

File: src/custom-element.js [hunk 12]
```diff
@@ -231,6 +231,9 @@ export function applyLayout_(element) {
     // Do nothing. Elements themselves will check whether the supplied
     // layout value is acceptable. In particular container is only OK
     // sometimes.
+  } else if (layout == Layout.FLEX_ITEM) {
+    // Do nothing.
+    // Layout will be determined by parent element with style 'display : flex'.
```
Comment (dvoytenko) [inline src/custom-element.js]: 'display : flex' -> `display: flex` with code-style quotes.

File: test/functional/test-layout.js [hunk 13]
```diff
@@ -279,6 +279,16 @@ describe('Layout', () => {
     expect(div.children.length).to.equal(0);
   });
 
+  it('layout=flex', () => {
```
Comment (dvoytenko) [inline test/functional/test-layout.js]: "layout=flex-item"

**Comment by dvoytenko** [comment]
Few more comments, but otherwise close to LG on my side.

File: css/amp.css [hunk 14]
```diff
@@ -78,6 +78,14 @@ body {
...
+  position: relative;
+  flex: 1 1 0;
+  align-self: stretch;
+  height: 100%;
```
Comment (dvoytenko) [inline css/amp.css]: Did you confirm if this always works as we expected? E.g. `flex-basis` takes over if parent is indeed `display: flex`?

File: src/custom-element.js (hunk 12 above)
Comment (zhouyx) [inline src/custom-element.js]: parent element with style "display: flex"
Like this? @dvoytenko

**Comment by zhouyx** [comment]
@dvoytenko  PTAL, Please give your final thoughts

File: css/amp.css (hunk 14 above)
Comment (zhouyx) [inline css/amp.css]: Tried add flex-basis. Works as long as having set `flex: 1 1 0`;

File: examples/layout-flex-item.amp.html [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+<h3>Flex-item in column</h3>
+<div class="container-col">
+   <amp-img id="child4" src="https://lh3.googleusercontent.com/pSECrJ82R7-AqeBCOEPGPM9iG9OEIQ_QXcbubWIOdkY=w400-h300-no-n" layout=flex-item></amp-img>
+  <amp-img id="child5" src="https://lh3.googleusercontent.com/5rcQ32ml8E5ONp9f9-Rf78IofLb9QjS5_0mqsY1zEFc=w400-h300-no-n" layout=flex-item></amp-img>
+  <amp-img id="child6" src="https://lh3.googleusercontent.com/Z4gtm5Bkxyv21Z2PtbTf95Clb9AE4VTR6olbBKYrenM=w400-h300-no-n" layout=flex-item></amp-img>
+</div>
+</body>
+</html>
+
+<h3>Element without layout typle 'flex-item', have to set children's height and width manually</h3>
```
Comment (dvoytenko) [inline examples/layout-flex-item.amp.html]: This markup should probably go before `</body>`. Right?

File: spec/amp-html-layout.md [hunk 16]
```diff
@@ -68,6 +68,8 @@ the layout of a `fill` element matches its parent.
 
 The element lets its children to define its size, much like a normal HTML `div`.
 
+### `flex-item`
+The element and other elements with layout type `flex-item` take their parent container's remaining space when parent has display : flex.
```
Comment (dvoytenko) [inline spec/amp-html-layout.md]: 1. What are "and other elements"?
2. "display : flex" -> `display: flex` with apostrophe quotes.

**Comment by dvoytenko** [comment]
Looks good. Couple minor fixes and it'll be ready for merge.

File: examples/layout-flex-item.amp.html (hunk 15 above)
Comment (zhouyx) [inline examples/layout-flex-item.amp.html]: Done

File: spec/amp-html-layout.md (hunk 16 above)
Comment (zhouyx) [inline spec/amp-html-layout.md]: Change to: The element and other elements in its parent

File: css/amp.css [hunk 17]
```diff
@@ -78,6 +78,15 @@ body {
   right: 0;
 }
 
+.-amp-layout-flex-item {
+  display: block;
+  position: relative;
+  flex: 1 1 0;
```
Comment (dvoytenko) [inline css/amp.css]: Ok, let me rephrase this. We need to confirm these properties:

1. `flex` already includes `flex-basis`. In our case it's 0. Should `flex` be actually be `flex: 1 1 auto`?
2. Why do we need `align-self: stretch`?
3. Do we even need width and height `100%` here? What happens without them? My initial idea behind them is to break layout when parent is NOT `display: flex`. I think it works and doesn't interfere when parent is as expected `display:flex`. But we need to make sure it works as expected.

Comment (zhouyx) [inline css/amp.css]: tested on property 1 and 2, made the change. 
For 3, the width and height `100%` works as expected. 
If the parent is `display:flex` setting width&height to `100%` won't be necessary. 
If parent is NOT `display:flex` children will need a width/height. So do we want the width and height `100%` in css in case parent is NOT `display:flex`?

**Comment by zhouyx** [comment]
@dvoytenko PTAL

File: test/functional/test-layout.js [hunk 18]
```diff
@@ -279,6 +279,16 @@ describe('Layout', () => {
     expect(div.children.length).to.equal(0);
   });
 
+  it('layout=flex-item', () => {
```
Comment (dvoytenko) [inline test/functional/test-layout.js]: We need another test (or modify this one) to show that width/height are set.

File: src/custom-element.js [hunk 19]
```diff
@@ -230,6 +230,15 @@ export function applyLayout_(element) {
     // Do nothing. Elements themselves will check whether the supplied
     // layout value is acceptable. In particular container is only OK
     // sometimes.
+  } else if (layout == Layout.FLEX_ITEM) {
+    // Set height and width to a flex item if they exist.
+    // The size set to a flex item could be overide by `display: flex` later.
```
Comment (dvoytenko) [inline src/custom-element.js]: "could be overide" -> "could be overridden"

**Comment by dvoytenko** [comment]
@zhouyx Great! Couple more cleanup/test requests.

File: test/functional/test-layout.js (hunk 18 above)
Comment (zhouyx) [inline test/functional/test-layout.js]: @dvoytenko  Done

File: src/custom-element.js (hunk 19 above)
Comment (zhouyx) [inline src/custom-element.js]: @dvoytenko  Done

**Comment by dvoytenko** [comment]
w00t! LGTM!

**Comment by zhouyx** [comment]
Woohoo! My first LGTM!
